src/analyses/cfg_dominators.h

```diff
@@ -172,58 +208,40 @@ void cfg_dominators_templatet<P, T, post_dom>::fixedpoint(P &program)
 
     bool changed=false;
     typename cfgt::nodet &node = cfg.get_node(current);
-    if(node.dominators.empty())
-    {
-      for(const auto &edge : (post_dom ? node.out : node.in))
-        if(!cfg[edge.first].dominators.empty())
-        {
-          node.dominators=cfg[edge.first].dominators;
-          node.dominators.insert(current);
-          changed=true;
-        }
-    }
+
+    auto potential_dominator = node.dominator;
 
     // compute intersection of predecessors
     for(const auto &edge : (post_dom ? node.out : node.in))
     {
-      const target_sett &other=cfg[edge.first].dominators;
-      if(other.empty())
+      const typename cfgt::nodet &other = cfg[edge.first];
+      if(!other.dominator)
         continue;
 
-      typename target_sett::const_iterator n_it=node.dominators.begin();
-      typename target_sett::const_iterator o_it=other.begin();
+      if (!potential_dominator)
+        potential_dominator = other.dominator;
 
-      // in-place intersection. not safe to use set_intersect
-      while(n_it!=node.dominators.end() && o_it!=other.end())
+      auto edge_index = *other.dominator;
+      while(potential_dominator != edge_index)
```

Comment:
This is the `intersect` function from the Cooper paper -- pull it out into a static function to make the correspondence with the paper clearer